doc/api/esm.md

```diff
@@ -1066,54 +1058,272 @@ const baseURL = pathToFileURL(process.cwd()).href;
     throw new Error(
       `imports must begin with '/', './', or '../'; '${specifier}' does not`);
   }
-  const resolved = new URL(specifier, parentModuleURL);
+  const resolved = new URL(specifier, parentURL);
   const ext = path.extname(resolved.pathname);
   if (!JS_EXTENSIONS.has(ext)) {
     throw new Error(
       `Cannot load file with non-JavaScript file extension ${ext}.`);
   }
   return {
-    url: resolved.href,
-    format: 'module'
+    url: resolved.href
   };
 }
 ```
 
 With this loader, running:
 
 ```console
-NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node x.js
+NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node main.js
 ```
 
-would load the module `x.js` as an ES module with relative resolution support
+would load the module `main.js` as an ES module with relative resolution support
 (with `node_modules` loading skipped in this example).
 
+### Get format hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+The `getFormat` hook provides a way to define a custom method of determining how
+a URL should be interpreted. This can be one of the following:
+
+| `format` | Description |
+| --- | --- |
+| `'builtin'` | Load a Node.js builtin module |
+| `'commonjs'` | Load a Node.js CommonJS module |
+| `'dynamic'` | Use a [dynamic instantiate hook][] |
+| `'json'` | Load a JSON file |
+| `'module'` | Load a standard JavaScript module (ES module) |
+| `'wasm'` | Load a WebAssembly module |
+
+```js
+import { request } from 'https';
+
+const formatMap = {
+  'text/javascript': 'module',
+  'application/javascript': 'module'
+};
+
+/**
+ * @param {string} url
+ * @param {object} context (currently empty)
+ * @param {function} defaultGetFormat
+ * @returns {object} response
+ * @returns {string} response.format
+ */
+export async function getFormat(url, context, defaultGetFormat) {
+  if (url.startsWith('https://')) {
+    return new Promise((resolve, reject) => {
+      request(url, {
+        method: 'HEAD'
+      }, (err, res) => {
+        if (err) reject(err);
+        else resolve({ format: formatMap[res.headers['content-type']] });
+      });
+    });
+  }
+  return defaultGetFormat(url, context, defaultGetFormat);
+}
+```
+
+### Get source hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+The `getSource` hook provides a way to define a custom method for retrieving
+the source code of an ES module specifier. This would allow a loader to
+potentially avoid reading files from disk.
+
+```js
+import { get } from 'https';
+
+/**
+ * @param {string} url
+ * @param {object} context
+ * @param {string} context.format
+ * @param {function} defaultGetSource
+ * @returns {object} response
+ * @returns {string|buffer} response.source
+ */
+export async function getSource(url, { format }, defaultGetSource) {
```

Comment:
Ideally in the examples we should not be destructuring and always just using eg `context.format`.